java/spark/src/main/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReader.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.arrow.vector.types.pojo.Schema;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.execution.arrow.ArrowWriter;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.io.IOException;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A custom arrow reader that supports writes Spark internal rows while reading data in batches.
+ */
+public class InternalRowWriterArrowReader extends ArrowReader {
+  private final Schema schema;
+  private final int batchSize;
+  private final Object monitor = new Object();
+  @GuardedBy("monitor")
+  private final Queue<InternalRow> rowQueue = new ConcurrentLinkedQueue<>();
+  @GuardedBy("monitor")
+  private volatile boolean finished;
+
+  private final AtomicLong totalBytesRead = new AtomicLong();
+  private ArrowWriter arrowWriter = null;
+
+  public InternalRowWriterArrowReader(BufferAllocator allocator, Schema schema, int batchSize) {
+    super(allocator);
+    Preconditions.checkNotNull(schema);
+    Preconditions.checkArgument(batchSize > 0);
+    this.schema = schema;
+    // TODO(lu) batch size as config?
+    this.batchSize = batchSize;
+  }
+
+  public void write(InternalRow row) {
+    Preconditions.checkNotNull(row);
+    synchronized (monitor) {
+      // TODO(lu) wait if too much elements in rowQueue
+      rowQueue.offer(row);
+      monitor.notify();
+    }
+  }
+
+  public void setFinished() {
+    synchronized (monitor) {
+      finished = true;
+      monitor.notify();
+    }
+  }
+
+  @Override
+  protected void prepareLoadNextBatch() throws IOException {
+    super.prepareLoadNextBatch();
+    // Do not use ArrowWriter.reset since it does not work well with Arrow JNI
+    arrowWriter = ArrowWriter.create(this.getVectorSchemaRoot());
+  }
+
+  @Override
+  public boolean loadNextBatch() throws IOException {
+    prepareLoadNextBatch();
+    int rowCount = 0;
+    synchronized (monitor) {
+      while (rowCount < batchSize) {
+        while (rowQueue.isEmpty() && !finished) {
+          try {
+            monitor.wait();
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new IOException("Interrupted while waiting for data", e);
+          }
+        }
+        if (rowQueue.isEmpty() && finished) {
+          break;
+        }
+        InternalRow row = rowQueue.poll();
+        if (row != null) {
+          arrowWriter.write(row);
+          rowCount++;
+        }
+      }
+    }
+    if (rowCount == 0) {
+      return false;
+    }
+    arrowWriter.finish();
+    // Calculate bytes read for the current record batch
+    // If the following code impacts performance, can be removed
+    VectorSchemaRoot root = this.getVectorSchemaRoot();
+    VectorUnloader unloader = new VectorUnloader(root);
+    try (ArrowRecordBatch recordBatch = unloader.getRecordBatch()) {
```

Comment:
A `RecordBatch` is just consumed to update the `totalBytesRead`.

Reply:
Yeah, there is no better way to compute bytes read. Alternative approach is to let bytesRead() throws unsupportedOperationException, in my testing, bytesRead is never called